policy-controller/k8s/index/src/server_authorization.rs

```diff
@@ -1,234 +1,91 @@
-            }
-        })
-    }
-
-    /// Updates the authorization and server indexes with a new or updated authorization instance.
-    fn apply(
-        &mut self,
-        authz: policy::ServerAuthorization,
-        servers: &mut SrvIndex,
-        cluster: &ClusterInfo,
-    ) {
-        let name = authz.name();
-        let authz = match mk_server_authz(authz, cluster) {
-            Ok(authz) => authz,
+        let authz = match client_authz(saz.spec.client, &namespace, self.cluster_info()) {
+            Ok(ca) => ca,
             Err(error) => {
-                warn!(saz = %name, %error);
+                tracing::warn!(%error, %namespace, saz = %name, "invalid authorization");
                 return;
             }
         };
 
-        match self.index.entry(name) {
-            Entry::Vacant(entry) => {
-                servers.add_server_authz(entry.key(), &authz.servers, authz.clients.clone());
-                entry.insert(authz);
-            }
+        self.ns_or_default(namespace).apply_server_authorization(
+            name,
+            server_selector(saz.spec.server),
+            authz,
+        )
+    }
 
-            Entry::Occupied(mut entry) => {
-                // If the authorization changed materially, then update it in all servers.
-                if entry.get() != &authz {
-                    servers.add_server_authz(entry.key(), &authz.servers, authz.clients.clone());
-                    entry.insert(authz);
-                }
+    fn delete(&mut self, namespace: String, name: String) {
+        if let Entry::Occupied(mut entry) = self.entry(namespace) {
+            entry.get_mut().delete_server_authorization(&*name);
+            if entry.get().is_empty() {
+                entry.remove();
             }
         }
     }
 
-    fn delete(&mut self, name: &str) {
-        self.index.remove(name);
-        debug!("Removed authz");
+    fn snapshot_keys(&self) -> HashMap<String, HashSet<String>> {
+        self.snapshot_server_authorizations()
     }
 }
 
-fn mk_server_authz(
-    srv: policy::server_authorization::ServerAuthorization,
-    cluster: &ClusterInfo,
-) -> Result<Authz> {
-    let policy::server_authorization::ServerAuthorization { metadata, spec, .. } = srv;
-
-    let servers = {
-        let policy::server_authorization::Server { name, selector } = spec.server;
-        match (name, selector) {
-            (Some(n), None) => ServerSelector::Name(n),
-            (None, Some(sel)) => ServerSelector::Selector(sel.into()),
-            (Some(_), Some(_)) => bail!("authorization selection is ambiguous"),
-            (None, None) => bail!("authorization selects no servers"),
-        }
-    };
+fn server_selector(s: k8s::policy::server_authorization::Server) -> ServerSelector {
+    s.name.map(ServerSelector::Name).unwrap_or_else(|| {
+        s.selector
+            .map(ServerSelector::Selector)
+            .unwrap_or_else(|| ServerSelector::Selector(k8s::labels::Selector::default()))
```

Comment:
nit: i think this could be simplified to just
```suggestion
       ServerSelector::Selector(s.selector.unwrap_or_default())
```